### PR TITLE
[AMD] [GLM5] Fuse shared-expert append into aiter grouped-topk (skip per-layer append kernel)

### PR DESCRIPTION
**Pull request by Jacob0226**
## Summary

- On GLM-5.2 (`GlmMoeDsaForCausalLM`) with the non-EP aiter grouped-topk MoE route, every decoder layer runs a separate `_fused_append_shared_experts` kernel to append the shared expert into the top-k ids/weights. Instead, pre-populate a **persistent** top-k buffer's shared-expert columns once (`id = num_experts + i`, `weight = fused_shared_experts_scaling_factor`) and let the aiter kernel write only the routed columns via row stride, so the per-layer append kernel is removed. Applies to both prefill and decode.
- **Bit-identical** to the plain append: the shared column uses the same constant `fused_shared_experts_scaling_factor` the aiter append writes (threaded into `biased_grouped_topk_gpu`), so it is correct for any shared-expert scaling.
- Safe fallback: the persistent buffer is sized to the max prefill batch (chunked-prefill-size / max-prefill-tokens, capped). For token counts above it, and in `_post_process_topk_ids`, the plain append path is used (mirrored condition) — no crash on large prefill batches.
- Active only when non-EP (`moe_ep_size == 1`) and `num_fused_shared_experts > 0`. Auto-enabled (no env).

## Changes

| File | Change |
| --- | --- |
| layers/moe/topk.py | Pre-populate persistent top-k buffer's shared-expert columns once; aiter grouped-topk writes only routed columns via row stride; drop the per-layer `_fused_append_shared_experts` kernel; guarded fallback to the plain append for large prefill batches / non-applicable cases |

## Scope

GLM-5.2 (`GlmMoeDsaForCausalLM`) MoE on the non-EP aiter grouped-topk route (gfx950 / MI355X). Other archs, EP mode, and non-applicable routes fall back to the existing per-layer append and are unaffected.

## Test plan

**Accuracy (GSM8K, MI355X TP4, GLM-5.2-MXFP4):**

| | Baseline | This PR |
| --- | --- | --- |
| GSM8K | 0.931 | 0.933 |

Within margin of error. (Decode trace confirms `_fused_append_shared_experts` is gone while `grouped_topk_kernel` runs.)

**Performance (MI355X TP4, GLM-5.2-MXFP4, docker `rocm/sgl-dev:v0.5.14-rocm720-mi35x-20260708`):** token throughput per GPU (tok/s/gpu, higher better) and median TPOT (ms, lower better), baseline vs this PR.

**i1024 / o1024:**

| Concurrency | TPUT base | TPUT PR | Δ TPUT | TPOT base | TPOT PR | Δ TPOT |
| --- | --- | --- | --- | --- | --- | --- |
| 4  | 72.1  | 72.6  | +0.7% | 13.24 | 13.13 | −0.8% |
| 8  | 126.8 | 127.9 | +0.9% | 15.22 | 15.00 | −1.4% |
| 16 | 196.8 | 198.4 | +0.8% | 19.49 | 19.33 | −0.8% |
| 32 | 312.0 | 313.4 | +0.5% | 24.75 | 24.64 | −0.4% |
| 64 | 470.1 | 471.9 | +0.4% | 32.81 | 32.69 | −0.4% |

**i8192 / o1024:**

| Concurrency | TPUT base | TPUT PR | Δ TPUT | TPOT base | TPOT PR | Δ TPOT |
| --- | --- | --- | --- | --- | --- | --- |
| 4  | 64.0  | 64.3  | +0.4% | 14.61 | 14.51 | −0.7% |
| 8  | 105.6 | 105.0 | −0.6% | 17.86 | 17.84 | −0.1% |
| 16 | 151.1 | 151.4 | +0.2% | 24.65 | 24.60 | −0.2% |
| 32 | 216.9 | 217.9 | +0.5% | 35.10 | 34.84 | −0.7% |
| 64 | 282.0 | 280.5 | −0.6% | 54.25 | 54.44 | +0.4% |

```
python3 -m sglang.launch_server \
  --model amd/GLM-5.2-MXFP4 \
  --tp 4 \
  --trust-remote-code \
  --tool-call-parser glm47 \
  --reasoning-parser glm45 \
  --watchdog-timeout 1200 \
  --mem-fraction-static 0.85 \
  --kv-cache-dtype fp8_e4m3 \
  --disable-radix-cache \
  --model-loader-extra-config '{"enable_multithread_load": true, "num_threads": 32}' \
  --dsa-prefill-backend tilelang \
  --dsa-decode-backend tilelang \
  --tokenizer-worker-num 8 \
  --enable-aiter-allreduce-fusion
```

E2E impact is small (within run-to-run noise); the change removes one kernel launch per layer (3 kernels → fewer) and is bit-identical, so it is a clean simplification with a slight net-positive tendency at low concurrency.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #31984281516](https://github.com/sgl-project/sglang/actions/runs/31984281516)
Latest PR Test (Extra): :x: [Run #31984281422](https://github.com/sgl-project/sglang/actions/runs/31984281422)

### CODE REVIEW

**Review by clintg6 — APPROVED** [review]
@Jacob0226 @sogalin @HaiShaw I have finished reviewing and testing the PR and found no impact on accuracy (95.3%) and 1% uplift on TPOT/TPUT for ISL/OSL 1k/1k.  I also tested gemini's suggestions (removing the try/except guard and replacing the mirrored condition with a shape check) and found no issues, but will leave it up to you whether to apply them.

File: python/sglang/srt/layers/moe/topk.py [hunk 1]
```diff
@@ -141,6 +141,69 @@ def routing(
...
+        sa = get_global_server_args()
+        cps = getattr(sa, "chunked_prefill_size", None) or 0
+        mpt = getattr(sa, "max_prefill_tokens", None) or 0
+        m = max(int(cps), int(mpt), 8192)  # 8192 floor for tiny configs
+        if int(cps) <= 0 and int(mpt) <= 0:
+            # chunked prefill disabled -> use the safety cap
+            m = _AITER_TOPK_FUSE_SHARED_MAX_TOKENS_CAP
+        _aiter_topk_fuse_shared_max_tokens_cache = min(
+            m, _AITER_TOPK_FUSE_SHARED_MAX_TOKENS_CAP
+        )
```
[the remark this answers is not in the document]
Comment (Jacob0226) [reply python/sglang/srt/layers/moe/topk.py:198]: Solved in commit abb96d9

File: python/sglang/srt/layers/moe/topk.py [hunk 2]
```diff
@@ -1903,25 +2000,39 @@ def _post_process_topk_ids(
...
-            else fused_shared_experts_scaling_factor
+        # Must mirror biased_grouped_topk_gpu's _shared_fuse condition EXACTLY
+        # (incl. the token<=MAX buffer bound): when the fast path fell back to the
+        # plain buffer (e.g. token > MAX during a large prefill), the shared experts
+        # were NOT pre-populated, so we must append them here.
+        _shared_fused_in_topk = (
+            num_fused_shared_experts > 0
+            and get_parallel().moe_ep_size == 1
+            and router_logits.shape[0] <= _get_aiter_topk_fuse_shared_max_tokens()
         )
```
[the remark this answers is not in the document]
Comment (Jacob0226) [reply python/sglang/srt/layers/moe/topk.py]: Applied the review suggestion: replaced the mirrored _shared_fuse condition in _post_process_topk_ids with a robust topk_ids.shape[1] == topk_config.top_k shape check.

**Review by sogalin — APPROVED** [review]

**Comment by Jacob0226** [comment]
@amd-bot ci-status

**Comment by Jacob0226** [comment]
image
CUDA CI (PR Test Base) ALL green

**Comment by Jacob0226** [comment]
image
AMD CI all green

File: python/sglang/srt/layers/moe/topk.py [hunk 3]
```diff
@@ -1728,6 +1831,9 @@ def biased_grouped_topk_cpu(
     num_fused_shared_experts: int = 0,
     routed_scaling_factor: Optional[float] = None,
     apply_routed_scaling_factor_on_output: Optional[bool] = False,
+    # Accepted for API parity with the GPU variant (aiter shared-expert fusion
+    # fast path); unused on CPU.
+    fused_shared_experts_scaling_factor: Optional[float] = None,
```
Comment (HaiShaw) [inline python/sglang/srt/layers/moe/topk.py]: Please remove this unused.

**Review by HaiShaw — COMMENTED** [review]
CI signal green. Nit: remove unused param.

**Review by HaiShaw — APPROVED** [review]

**Comment by hzh0425** [comment]
It seems that this PR broke CI

https://github.com/sgl-project/sglang/actions/runs/31984281516/job/95256531367
